contracts/fio.perms/fio.perms.cpp

```diff
@@ -0,0 +1,477 @@
+            auto fee_iter = fees_by_endpoint.find(endpoint_hash);
+            fio_400_assert(fee_iter != fees_by_endpoint.end(), "endpoint_name", ADD_PERMISSION_ENDPOINT,
+                           "FIO fee not found for endpoint", ErrorNoEndpoint);
+            const uint64_t fee_amount = fee_iter->suf_amount;
+            const uint64_t fee_type = fee_iter->type;
+
+            fio_400_assert(fee_type == 0, "fee_type", to_string(fee_type),
+                           "unexpected fee type for endpoint transfer_fio_domain, expected 0",
+                           ErrorNoEndpoint);
+
+            fio_400_assert(max_fee >= (int64_t) fee_amount, "max_fee", to_string(max_fee),
+                           "Fee exceeds supplied maximum.",
+                           ErrorMaxFeeExceeded);
+
+            fio_fees(actor, asset(fee_amount, FIOSYMBOL), ADD_PERMISSION_ENDPOINT);
+            processbucketrewards(tpid, fee_amount, get_self(), actor);
+
+            if (fee_amount > 0) {
+                INLINE_ACTION_SENDER(eosiosystem::system_contract, updatepower)
+                        (SYSTEMACCOUNT, {{_self, "active"_n}},
+                         {actor, true}
+                        );
+            }
+
+            //ram bump
+            if (ADDPERMISSIONRAM > 0) {
+                action(
+                        permission_level{SYSTEMACCOUNT, "active"_n},
+                        "eosio"_n,
+                        "incram"_n,
+                        std::make_tuple(actor, ADDPERMISSIONRAMBASE + (ADDPERMISSIONRAM * permission_info.size()))
+                ).send();
+            }
+            const string response_string = "{\"status\": \"OK\", \"fee_collected\" : "+ to_string(fee_amount) +"}";
+            send_response(response_string.c_str());
+
+        }
+
+        [[eosio::action]]
+        void
+        remperm(const name &grantee_account,
+                const string &permission_name,
+                const string &object_name,
+                const int64_t &max_fee,
+                const string &tpid,
+                const name &actor
+        ) {
+
+
+            print("remperm --      called. \n");
+
+            require_auth(actor);
+
+
+            fio_400_assert(permission_name.length()  > 0, "permission_name", permission_name,
+                           "Permission name is invalid", ErrorInvalidPermissionName);
+
+            string useperm = makeLowerCase(permission_name);
+
+            // error if permission name is not the expected name register_address_on_domain.
+            fio_400_assert(useperm.compare(REGISTER_ADDRESS_ON_DOMAIN_PERMISSION_NAME) == 0, "permission_name", permission_name,
+                           "Permission name is invalid", ErrorInvalidPermissionName);
+
+             // error if object name is not * or is not in the domains table
+            fio_400_assert(object_name.size()  > 0, "object_name", object_name,
+                           "Object name is invalid", ErrorInvalidObjectName);
```

Comment:
Same as above. Generic "object name invalid" errors should be replaced with messages that will make it clear what is wrong with the transaction.